### PR TITLE
feat[lang]!: make `@external` modifier optional in `.vyi` files

### PR DESCRIPTION
**Pull request by cyberthirst**
### What I did
- as per https://github.com/vyperlang/vyper/issues/4169#issuecomment-2203672462: make `@external` visibility in `vyi` files optional
- also should fix: https://github.com/vyperlang/vyper/issues/4169

### How to verify it
- added some tests

### Commit Message
```
make `@external` visibility in `.vyi` files optional.

additionally, fix a panic when functions in an interface have the wrong
visibility
```

edit: linked wrong issue

### CODE REVIEW

File: vyper/semantics/types/function.py [hunk 1]
```diff
@@ -752,7 +767,10 @@ def _parse_decorators(
             raise StructureException("Bad decorator syntax", decorator)
 
     if function_visibility is None:
-        function_visibility = FunctionVisibility.INTERNAL
+        if is_interface:
```
Comment (charles-cooper) [inline vyper/semantics/types/function.py:770]: i think we should actually bubble up `None` for function_visibility and let the caller handle it

File: vyper/semantics/types/module.py [hunk 2]
```diff
@@ -117,8 +117,9 @@ def _is_function_implemented(fn_name, fn_type):
 
             to_compare = fns_by_name[fn_name]
             assert isinstance(to_compare, ContractFunctionT)
+            assert isinstance(fn_type, ContractFunctionT)
 
-            return to_compare.implements(fn_type)
+            return fn_type.implements(to_compare)
```
Comment (charles-cooper) [inline vyper/semantics/types/module.py:122]: seems wrong! compare_type is not commutative

Comment (cyberthirst) [reply vyper/semantics/types/module.py]: reverted back

File: vyper/semantics/types/function.py [hunk 3]
```diff
@@ -327,10 +327,15 @@ def from_vyi(cls, funcdef: vy_ast.FunctionDef) -> "ContractFunctionT":
...
+        function_visibility, state_mutability, nonreentrant = _parse_decorators(
+            funcdef, is_interface=True
+        )
 
-        if nonreentrant:
-            raise FunctionDeclarationException("`@nonreentrant` not allowed in interfaces", funcdef)
+        assert not nonreentrant
+
+        # it's redundant to specify visibility in vyi - always should be external
+        function_visibility = function_visibility or FunctionVisibility.EXTERNAL
```
Comment (charles-cooper) [inline vyper/semantics/types/function.py:337]: more explicit:
```suggestion
        if function_visibility is None:
            FunctionVisibility.EXTERNAL
```

File: vyper/semantics/types/function.py [hunk 4]
```diff
@@ -377,6 +382,9 @@ def from_FunctionDef(cls, funcdef: vy_ast.FunctionDef) -> "ContractFunctionT":
         """
         function_visibility, state_mutability, nonreentrant = _parse_decorators(funcdef)
 
+        # it's redundant to specify internal visibility - it's implied by not being external
+        function_visibility = function_visibility or FunctionVisibility.INTERNAL
```
Comment (charles-cooper) [inline vyper/semantics/types/function.py:386]:
```suggestion
        if function_visibility is None:
            FunctionVisibility.INTERNAL
```

File: vyper/semantics/types/function.py (hunk 3 above)
Comment (cyberthirst) [reply vyper/semantics/types/function.py]: addressed in: a6f1a86ae7f3a6c920682e1b30c8675244b4a504

File: vyper/semantics/types/function.py (hunk 4 above)
Comment (cyberthirst) [reply vyper/semantics/types/function.py]: addressed in: a6f1a86ae7f3a6c920682e1b30c8675244b4a504

File: vyper/semantics/types/function.py (hunk 1 above)
Comment (cyberthirst) [reply vyper/semantics/types/function.py]: addressed in: 1ee63b6758b800af4ab63cce7176f725bfa9f7eb

File: vyper/semantics/types/function.py [hunk 5]
```diff
@@ -714,6 +728,10 @@ def _parse_decorators(
         if decorator.get("id") == "nonreentrant":
             if nonreentrant_node is not None:
                 raise StructureException("nonreentrant decorator is already set", nonreentrant_node)
+            if is_interface:
```
Comment (charles-cooper) [inline vyper/semantics/types/function.py:731]: i think we can move this validation up to the caller

File: vyper/semantics/types/function.py [hunk 6]
```diff
@@ -695,8 +709,8 @@ def _parse_return_type(funcdef: vy_ast.FunctionDef) -> Optional[VyperType]:
 
 
 def _parse_decorators(
-    funcdef: vy_ast.FunctionDef,
-) -> tuple[FunctionVisibility, StateMutability, bool]:
+    funcdef: vy_ast.FunctionDef, is_interface: bool = False
```
Comment (charles-cooper) [inline vyper/semantics/types/function.py:712]: not sure we need this `is_interface` kwarg

File: vyper/semantics/types/module.py [hunk 7]
```diff
@@ -107,7 +107,8 @@ def _ctor_modifiability_for_call(self, node: vy_ast.Call, modifiability: Modifia
     def validate_implements(
         self, node: vy_ast.ImplementsDecl, functions: dict[ContractFunctionT, vy_ast.VyperNode]
     ) -> None:
-        fns_by_name = {fn_t.name: fn_t for fn_t in functions.keys()}
+        # only external functions can implement interfaces
+        fns_by_name = {fn_t.name: fn_t for fn_t in functions.keys() if fn_t.is_external}
```
Comment (charles-cooper) [inline vyper/semantics/types/module.py:111]: we should assert that all the functions are external, but leave it to the caller to filter

File: vyper/semantics/analysis/module.py [hunk 8]
```diff
@@ -414,7 +414,7 @@ def visit_ImplementsDecl(self, node):
             raise StructureException(msg, node.annotation, hint=hint)
 
         # grab exposed functions
-        funcs = self._exposed_functions
+        funcs = self._all_functions
```
Comment (charles-cooper) [inline vyper/semantics/analysis/module.py:417]: i think `funcs = [fn_t for fn_t in self._all_functions if fn_t.is_external]`

File: vyper/semantics/types/function.py (hunk 6 above)
Comment (cyberthirst) [reply vyper/semantics/types/function.py]: lifted the interface-dependant checks in: https://github.com/vyperlang/vyper/pull/4178/commits/b8389641c6020b2b78728e4f78e6b663450601e1

File: vyper/semantics/types/function.py (hunk 5 above)
Comment (cyberthirst) [reply vyper/semantics/types/function.py]: lifted all non-general checks to callers in: https://github.com/vyperlang/vyper/pull/4178/commits/b8389641c6020b2b78728e4f78e6b663450601e1

File: vyper/semantics/analysis/module.py (hunk 8 above)
Comment (cyberthirst) [reply vyper/semantics/analysis/module.py]: addressed in: https://github.com/vyperlang/vyper/pull/4178/commits/e4f6622cd729eb7cdd26e6c8e8c455082f4500ad

File: vyper/semantics/types/module.py (hunk 7 above)
Comment (cyberthirst) [reply vyper/semantics/types/module.py]: addressed in: https://github.com/vyperlang/vyper/pull/4178/commits/e4f6622cd729eb7cdd26e6c8e8c455082f4500ad

File: vyper/semantics/types/function.py [hunk 9]
```diff
@@ -330,7 +330,22 @@ def from_vyi(cls, funcdef: vy_ast.FunctionDef) -> "ContractFunctionT":
         function_visibility, state_mutability, nonreentrant = _parse_decorators(funcdef)
 
         if nonreentrant:
-            raise FunctionDeclarationException("`@nonreentrant` not allowed in interfaces", funcdef)
+            decorator = next(d for d in funcdef.decorator_list if d.id == "nonreentrant")
```
Comment (charles-cooper) [inline vyper/semantics/types/function.py:333]: maybe `_parse_decorators` should actually return more structured output, so we don't need to do these searches (which is duplicating code with `_parse_decorators`)

File: vyper/semantics/types/function.py [hunk 10]
```diff
@@ -507,7 +530,10 @@ def implements(self, other: "ContractFunctionT") -> bool:
         if return_type and not return_type.compare_type(other_return_type):  # type: ignore
             return False
 
-        return self.mutability == other.mutability
+        if self.mutability != other.mutability:
+            return False
+
+        return self.visibility == other.visibility
```
Comment (charles-cooper) [inline vyper/semantics/types/function.py:536]: i think returning False here is unreachable, not sure why this change is needed

File: vyper/semantics/types/module.py [hunk 11]
```diff
@@ -116,7 +117,8 @@ def _is_function_implemented(fn_name, fn_type):
                 return False
 
             to_compare = fns_by_name[fn_name]
-            assert isinstance(to_compare, ContractFunctionT)
+            assert to_compare.is_external
+            assert all(isinstance(f, ContractFunctionT) for f in [to_compare, fn_type])
```
Comment (charles-cooper) [inline vyper/semantics/types/module.py:121]: nit:
```suggestion
            assert isinstance(to_compare, ContractFunctionT)
            assert isinstance(fn_type, ContractFunctionT)
```

File: vyper/semantics/types/function.py (hunk 10 above)
Comment (cyberthirst) [reply vyper/semantics/types/function.py:536]: currently this should be true. added the case for the sake of future-proofing the func in case the caller's semantics change / new callers are added

File: vyper/semantics/types/module.py (hunk 11 above)
Comment (cyberthirst) [reply vyper/semantics/types/module.py]: addressed in: https://github.com/vyperlang/vyper/pull/4178/commits/878b966b8133c47f80a2e8c894512e929fdd75a1

File: vyper/semantics/types/function.py (hunk 9 above)
Comment (cyberthirst) [reply vyper/semantics/types/function.py:333]: i think the current implementation is fairly simple and easy to refactor in the future. i favor avoiding further changes for this lightweight release

Comment (charles-cooper) [reply vyper/semantics/types/function.py:333]: add a comment then -- something along the lines of
```suggestion
            # TODO: refactor so parse_decorators returns the AST location
            decorator = next(d for d in funcdef.decorator_list if d.id == "nonreentrant")
```

File: vyper/semantics/types/function.py (hunk 10 above)
Comment (charles-cooper) [reply vyper/semantics/types/function.py:536]: hmm. in that case we should have an assertion, to communicate to the reader what the assumption is, not leave the reader combing the codebase trying to find cases where the two visibilities are different.

File: vyper/semantics/types/function.py [hunk 12]
```diff
@@ -700,7 +726,7 @@ def _parse_return_type(funcdef: vy_ast.FunctionDef) -> Optional[VyperType]:
 
 def _parse_decorators(
     funcdef: vy_ast.FunctionDef,
-) -> tuple[FunctionVisibility, StateMutability, bool]:
+) -> tuple[FunctionVisibility | None, StateMutability, bool]:
```
Comment (charles-cooper) [inline vyper/semantics/types/function.py:729]: nit: prefer `Optional[FunctionVisibility]`

Comment (cyberthirst) [reply vyper/semantics/types/function.py]: addressed in: https://github.com/vyperlang/vyper/pull/4178/commits/e50e44e3d39e9e1d3d525a4922f56589f9e6bfbc

File: vyper/semantics/types/function.py (hunk 10 above)
Comment (cyberthirst) [reply vyper/semantics/types/function.py]: addressed in: https://github.com/vyperlang/vyper/pull/4178/commits/1e3541d83b8d0f70410751cec7bdf3ac51dac6a1

**Review by charles-cooper — APPROVED** [review]
thanks!

File: vyper/semantics/types/function.py [hunk 13]
```diff
@@ -330,7 +330,23 @@ def from_vyi(cls, funcdef: vy_ast.FunctionDef) -> "ContractFunctionT":
...
+        # it's redundant to specify visibility in vyi - always should be external
+        if function_visibility is None:
+            function_visibility = FunctionVisibility.EXTERNAL
+
+        if function_visibility != FunctionVisibility.EXTERNAL:
+            nonexternal = next(
+                d for d in funcdef.decorator_list if d.id in FunctionVisibility.values()
+            )
+            raise FunctionDeclarationException(
+                "Interface functions can only be marked as `@external`", nonexternal
```
[the remark this answers is not in the document]
Comment (cyberthirst) [reply vyper/semantics/types/function.py:348]: hmm, we've improved the reporting a bit, so the message refers to the corresponding source construct

```vyper
@internal
@payable
def bar():
    ...
```

would yield:
```
vyper.exceptions.FunctionDeclarationException: Interface functions can only be marked as `@external`

  contract "tests/custom/i.vyi:1", function "bar", line 1:1 
  ---> 1 @internal
  --------^
       2 @payable
```

would you say that it's still confusing?

[the remark this answers is not in the document]
Comment (pcaversaccio) [reply vyper/semantics/types/function.py:348]: I personally think error messages should be as precise as possible and thus I still think we should mention somehow that it's a _visibility_ decorator.
